core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java

```diff
@@ -722,16 +720,25 @@ protected Query doToQuery(QueryShardContext context) throws IOException {
         if (query == null) {
             return null;
         }
+
+        //save the BoostQuery wrapped structure if present
+        List<Float> boosts = new ArrayList<>();
+        while(query instanceof BoostQuery) {
+            BoostQuery boostQuery = (BoostQuery) query;
+            boosts.add(boostQuery.getBoost());
+            query = boostQuery.getQuery();
+        }
+
         query = Queries.fixNegativeQueryIfNeeded(query);
         if (query instanceof BooleanQuery) {
             query = Queries.applyMinimumShouldMatch((BooleanQuery) query, this.minimumShouldMatch());
         }
-        return query;
-    }
 
-    @Override
-    protected void setFinalBoost(Query query) {
-        //we need to preserve the boost that came out of the parsing phase
-        query.setBoost(query.getBoost() * boost);
+        //restore the previous BoostQuery wrapping
+        for (int i = boosts.size() - 1; i >= 0; i--) {
+            query = new BoostQuery(query, boosts.get(i));
+        }
+
+        return query;
```

Comment:
I think that this unwrapping business is needed only in query_string as it's the only query that supports providing boosts as part of the query, thus the resulting boolean query might have been boosted multiple times. I think the same cannot happen with simple query string , multi match query and similar queries that call Queries#applyMinimumShouldMatch... but I might be missing something